src/Google.Bigquery.V2/InsertRow.cs

```diff
@@ -0,0 +1,138 @@
+// limitations under the License.
+
+using Google.Apis.Bigquery.v2.Data;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+
+namespace Google.Bigquery.V2
+{
+    /// <summary>
+    /// A row used for streaming data insertion.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// This is effectively a dictionary where each key is the field name (a string)
+    /// and the value is the corresponding value for that field.
+    /// Supported value types:
+    /// </para>
+    /// <list type="bullet">
+    ///   <item></item>
+    /// </list>
+    /// </remarks>
+    public sealed class InsertRow : IEnumerable
+    {
+        private static HashSet<Type> ValidTypes = new HashSet<Type>
+        {
+            typeof(int), typeof(long), typeof(uint),
+            typeof(ulong), typeof(short), typeof(ushort),
+            typeof(string), typeof(byte[]),
+            typeof(DateTime), typeof(DateTimeOffset)
+        };
+        /// <summary>
+        /// To help ensure data consistency, you can supply an <see cref="InsertId" /> for each inserted row.
+        /// BigQuery remembers this ID for at least one minute. If you try to stream the same set of rows within
+        /// that time period and the insertId property is set, BigQuery uses the property to de-duplicate
+        /// your data on a best effort basis.
+        /// </summary>
+        public string InsertId { get; set; }
+        private readonly IDictionary<string, object> _fields = new Dictionary<string, object>();
+
+        public InsertRow(string insertId)
+        {
+            InsertId = insertId;
+        }
+
+        public InsertRow()
+        {
+        }
+
+        public void Add(IDictionary<string, object> fields)
+        {
+            Preconditions.CheckNotNull(fields, nameof(fields));
+            foreach (var entry in fields)
+            {
+                Add(entry.Key, entry.Value);
+            }
+        }
+
+        public void Add(string key, object value)
+        {
+            TableSchemaBuilder.ValidateFieldName(key, nameof(key));
+            ValidateValue(value, nameof(value));
+            _fields.Add(key, value);
+        }
+
+        /// <summary>
+        /// Returns an enumerator over the fields in this insert row. This
+        /// method is only provided in order to implement <see cref="IEnumerable"/>,
+        /// to permit collection initializers.
+        /// </summary>
+        /// <returns>An enumerator over the fields in this insert row.</returns>
+        IEnumerator IEnumerable.GetEnumerator() => _fields.GetEnumerator();
+
+        public object this[string key]
+        {
+            get
+            {
+                TableSchemaBuilder.ValidateFieldName(key, nameof(key));
+                return _fields[key];
+            }
+            set
+            {
+                TableSchemaBuilder.ValidateFieldName(key, nameof(key));
```

Comment:
Nope, because they behave differently - `Add` throws if the key already exists; the indexer doesn't. (Admittedly if I'd written any documentation for this, it would be more obvious...)